RFC#0192 -  ensures workers do not get unnecessarily killed

JohanLorenzo opened the pull request.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,58 @@
+
+### What if we deploy new worker images?
+
+Long-lived workers will have to be killed if there's a change in their config, including
```
This goes along with the #191 proposal quite well. 
With introduction of launch configurations, workers would be able to query w-m periodically to see if config is still active, and restart/shutdown if not. 

We could probably create a more generic endpoint which would answer if worker should terminate:
- `yes` if config changed
- `no` if `minCapacity` is not observed.

However, I'm not totally sure how to solve this issue if all workers at once ask if they could shutdown, and all get `yes`.. just to have zero running workers after

JohanLorenzo replied:
> This goes along with the https://github.com/taskcluster/taskcluster-rfcs/pull/191 proposal quite well.

Nice! 👍 

> However, I'm not totally sure how to solve this issue if all workers at once ask if they could shutdown, and all get yes.. just to have zero running workers after

I guess that's an okay behavior, at least for a first iteration. Maybe `worker-manager` could remember how many workers are being shut down so it doesn't tell more workers to turn off.

lotas replied:
unfortunately that part was never implemented because different strategy existed - `deploymentId`

JohanLorenzo replied:
Resolving thread now that #191 is live in production.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,58 @@
+
+## When `minCapacity` is not yet met
+
+Here, `worker-manager` should increase `afterIdleSeconds` to a much higher value (e.g.:
```
"tweaking" idle seconds could be a good solution for existing workers that are unlikely to get upgraded soon.

There should be some upper limit of what worker-manager could set for such workers to balance `minCapacity` promise and avoiding additional costs from running them for too long

JohanLorenzo replied:
To me, if we set `minCapacity` then we expect to have a number of workers at any time of the day. I would expect these workers to keep running. If some worker-types shouldn't be running for too long, I wonder if we should set `minCapacity` to 0.

lotas replied:
that's true! Indeed, if someone sets min value he probably knows what he's doing, no need to restrict

djmitche replied:
Sorry term I think this isn't a problem. 

Long term something probabilistic could work: p(no) = 0.5 if exactly at minCapacity, p lower if higher than minCapacity

JohanLorenzo replied:
I wonder if we need a probabilistic model at this stage. To me, there are already several random factors impacting workers (e.g.: spot instance termination, network issues). So having a deterministic model here may help diagnosing what goes wrong with the implementation. I may be missing something in my understanding of the problem, though.

JohanLorenzo replied:
Resolving thread given that we now have a much simpler approach.

JohanLorenzo commented:
@lotas 👋 Do you see anything more we should discuss? If not, is okay to open up the discussion in the next community meeting?

lotas commented:
> @lotas 👋 Do you see anything more we should discuss? If not, is okay to open up the discussion in the next community meeting?

please bring it up during the next community meeting, would be best! Thanks

JohanLorenzo commented:
I took another stab at it now that https://github.com/taskcluster/taskcluster-rfcs/pull/191 is implemented 😃

matt-boris commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,456 @@
+
+**Current Implementation Issues:**
+- Worker-manager enforces minCapacity by spawning new workers when capacity drops below threshold
+- Generic-worker shuts down after `afterIdleSeconds` regardless of minCapacity requirements
```
```suggestion
- Generic-worker shuts down after `idleTimeoutSecs` regardless of minCapacity requirements
```

Something to note (as information for the entire proposal) is if this config is set to `0`, the worker will continue running indefinitely. I could see this, mixed with an on-demand instance type, nicely fill the void of _true_ min capacity workers.

JohanLorenzo replied:
TIL! I didn't know g-w had this capacity. @lotas, do you see any issues with `0` if w-m is able to terminate instances in the cloud?

JohanLorenzo replied:
Sorry, Github hadn't refreshed. I just saw the thread on https://github.com/taskcluster/taskcluster-rfcs/pull/192#discussion_r2382439689. Let's continue the conversation there.

matt-boris replied:
> TIL! I didn't know g-w had this capacity.

This use case is typically used on hardware workers :)

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,456 @@
+### Core Concept: MinCapacity Workers
+
+Workers fulfilling `minCapacity >= 1` requirements should:
+1. **Extended Idle Timeout**: Receive significantly longer idle timeouts to preserve caches
```
What would be the desired timeout? Should it be also part of worker-pool config?

matt-boris replied:
`idleTimeoutSecs` of `0` will continue running indefinitely, fwiw.

lotas replied:
I like this, this makes implementation much simpler. Min capacity workers get idle timeout of `0`

matt-boris replied:
Yeah and then we could play around with future changes of making worker manager use on-demand instance type too, as I laid out in https://github.com/taskcluster/taskcluster-rfcs/pull/192#pullrequestreview-3272281542.

But I think just overwriting the worker config `idleTimeoutSecs` to `0` would be a huge quick win.

JohanLorenzo replied:
Great! Let's do that, then. Thanks for suggesting it @matt-boris!

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,456 @@
+
+Workers fulfilling `minCapacity >= 1` requirements should:
+1. **Extended Idle Timeout**: Receive significantly longer idle timeouts to preserve caches
+2. **Priority Replacement**: Be immediately replaced when they fail or become unhealthy
```
challenge here would be to let worker manager track which workers were "promoted" to the long-living ones. And every time the worker is getting removed/killed, worker manager would need to scan all workers to know how many long-lived ones are still running .
If it would be tracked separately, then it would likely change the estimator and provisioner to treat them in a certain way.

I'm a bit concerned if this special treatment (immediate replacement)  is the right approach.

I think a better approach might be to not immediately replace it, but let all tasks drain, and as soon as there will be `totalRunning < minCapacity` - worker runner could provision this difference with the extended timeout workers. But not during `totalRunning >>> minCapacity` periods.

lotas replied:
I think we can achieve this behavior with the least changes, if we just keep existing loops as is, with the only difference being:
- if currently running < mincapacity -> launch one with idle timeout: `0`
- otherwise business as usual

This way during off-peak hours, short lived ones would die off, and w-m would launch new with infinite run time.

This looks like a reasonably simple change

JohanLorenzo replied:
Good call! I updated the RFC to highlight we just want to provision such workers when `totalRunning < minCapacity`.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,456 @@
+Workers fulfilling `minCapacity >= 1` requirements should:
+1. **Extended Idle Timeout**: Receive significantly longer idle timeouts to preserve caches
+2. **Priority Replacement**: Be immediately replaced when they fail or become unhealthy
+3. **Per-Variant MinCapacity**: Each pool variant (gecko-1 vs gecko-3) maintains its own minCapacity workers
```
this is already guaranteed at the moment - each worker pool has own min capacity

JohanLorenzo replied:
True. I got rid of this line as it's causing confusion.

matt-boris reviewed (commented):
To me, this problem is solved with minimal code changes to worker manager to allow for a secondary worker config to be used when provisioning workers to fill the `minCapacity` void. This allows us to just use what's already there.

This could look like allowing a second block of worker config in the worker pool config or even just have worker manager automatically edit the `idleTimeoutSecs` to `0` and make it an on-demand instance type for the workers they're provisioning to get the capacity to `minCapacity`.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,456 @@
+**Current Implementation Issues:**
+- Worker-manager enforces minCapacity by spawning new workers when capacity drops below threshold
+- Generic-worker shuts down after `afterIdleSeconds` regardless of minCapacity requirements
+- Gap exists between worker shutdown and replacement detection/provisioning
```
I don't think we would be able to get rid of this gap entirely, or it may appear and go away sporadically. 

If empty pool is being provisioned with `minCapacity` workers at once, and they all get same large `idleTimeout` - it would mean that they would stop around the same time. 
Of course, after few cycles, there would be long living ones that started at different times, but they would still be stopped eventually. And this gap will happen. 

Unless, we are talking about timeouts that are much larger than max delay between tasks (meaning worker would never be able to shut down)?  This will probably have some potential downside with free disk space issue?

matt-boris replied:
> This will probably have some potential downside with free disk space issue?

Probably wouldn't be an issue, especially with the multiuser engine. So all task users/directories would be cleaned up. As far as having lots of caches around, generic worker will automatically garbage collect these, starting with file caches (which are easy to redownload) and then will move onto the writable directory caches that are written to/read from during task execution. It performs this cleanup one cache at a time until there's `config.RequiredDiskSpaceMegabytes` of free space available.

JohanLorenzo replied:
(Edit) Striking out comment given that we're going to spawn workers with `idleTimeoutSecs` set to 0

> ~~If empty pool is being provisioned with minCapacity workers at once, and they all get same large idleTimeout - it would mean that they would stop around the same time.~~

~~If I understand correctly, that is true only in this condition:~~

~~All `minCapacity` worker take their last task at around the same time and these last tasks take the same time to be resolved. Otherwise, the randomness on task duration and on number of tasks queued at any time should make the gap less visible, right?~~

~~Moreover, at the moment, we only `minCapacity` to 1. We don't use any greater value. So I agree the gap will be seen. If we could make it occur less often, we improve the current situation.~~

> ~~Unless, we are talking about timeouts that are much larger than max delay between tasks~~

~~I'm not sure I follow. What is the max delay between tasks?~~

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,456 @@
+- Keyed `minCapacity: 1`: [`gecko-1/decision-gcp`, `gecko-3/decision-gcp`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L976), and [pools matching `(app-services|glean|mozillavpn|mobile|mozilla|translations)-1`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L1088)
+
+**Current Implementation Issues:**
+- Worker-manager enforces minCapacity by spawning new workers when capacity drops below threshold
```
One extra side-effect that we will have to handle is when someone would lower `minCapacity` for a pool. What do we do with those extra workers running forever? 

It can also be a mistake during deployment time, if somebody applies config by accident with a really high number here, we would create "forever-workers" without means to kill them automatically. This can probably be left for worker deployers to figure out how to kill the excess of workers? 

Just something to keep in mind, how should worker manager handle this (or ignore)

JohanLorenzo replied:
Resolving this thread per https://github.com/taskcluster/taskcluster-rfcs/pull/192#discussion_r2382877200

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,456 @@
+- Worker-manager determines worker role at spawn time based on current pool needs
+- Workers are launched with predetermined `intended_role`
+- Uses existing [`launch_config_id` system](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/data.js#L313-L362) to ensure per-variant capacity tracking
+- **MinCapacity Worker Selection**: When capacity planning, prioritize spawning minCapacity workers first
```
i think this is redundant. workers would be created as is to run tasks that are pending, it doesn't matter which order they would be launched in

JohanLorenzo replied:
Sorry, I'm not sure I fully understand your comment. If we prioritize `minCapacity` workers then we prevent w-m from spawning short-lived workers (which would then quickly timeout leaving `minCapacity` unfulfilled). What did I miss?

JohanLorenzo replied:
Resolving thread because we agreed on only provisioning permanent workers when we're the number of workers falls below `minCapacity` https://github.com/taskcluster/taskcluster-rfcs/pull/192#discussion_r2382446393

matt-boris replied:
From how I understand it, I think he's saying how worker manager shouldn't care of it's a "minCapacity" worker or not. If there's a 0 capacity in a pool with pending tasks, then of course worker manager will provision new workers to do the work. I think we could then _consider_ these workers a "minCapacity" worker so that worker manager knows to set the idle timeout to 0 (for each worker that brings capacity up to that `minCapacity` target).

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
````diff
@@ -0,0 +1,456 @@
+```yaml
+config:
+  minCapacity: 1
+  minCapacityIdleTimeoutSecs: 86400  # 24 hours - enables minCapacity workers
````
or it could be `0` as Boris suggested, but the risk will be much higher if someone changes minCapacity to a lower value - we might not be able to kill those "Forever-workers" easily

JohanLorenzo replied:
Let's move the conversation to https://github.com/taskcluster/taskcluster-rfcs/pull/192#discussion_r2382428174. Resolving thread.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
````diff
@@ -0,0 +1,456 @@
+**Provisioning Strategy:**
+```javascript
+// Enhanced estimator logic for lifecycle-based workers
+async estimateCapacity(pool) {
````
yay! :) let's ignore this part, I think it  is making already not simple estimator much more less simpler 🤣

JohanLorenzo replied:
I removed the code but I converted it to plain text English. Please let me know if you want me to get rid of the estimator part completely.

JohanLorenzo replied:
Closing thread. The logic with the estimator is much simpler after we decided only provisioning permanent workers when we're the number of workers falls below `minCapacity`

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
````diff
@@ -0,0 +1,456 @@
+**Capacity Reduction Strategy:**
+```javascript
+// Hybrid approach: minimize cache loss when reducing capacity
+async handleMinCapacityDecrease(pool, newMinCapacity) {
````
ah, here goes this hidden edge-case handling.

I think it should be part of the RFC description, not implementation suggestion

JohanLorenzo replied:
Understood, done!

JohanLorenzo commented:
Thank you very much @lotas and @matt-boris for your insightful input! I like where the RFC is going. It's now much simpler 👍 I believe I addressed all the comments and it's now ready for another round of review. Please let me know if there's anything I missed.

matt-boris commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+
+**Current Implementation Issues:**
+- Worker-manager enforces `minCapacity` by spawning new workers when capacity drops below threshold
+- Generic-worker shuts down after `afterIdleSeconds` regardless of `minCapacity` requirements
```
This is docker worker's specific config var, just highlighting for accuracy.


```suggestion
- Generic-worker shuts down after `idleTimeoutSecs` regardless of `minCapacity` requirements
```

matt-boris commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+
+## Compatibility Considerations
+
+**Backward Compatibility:**
+- Opt-in via `minCapacityIdleTimeoutSecs` configuration
+- Existing pools continue current behavior unless explicitly enabled
```
Honestly, I'd consider implementing this would be a "bug fix" to worker manager and not a new feature that we'd need to have backward compatibility for to the old, "broken" `minCapacity` functionality.

matt-boris replied:
Now that I think about this more, I think this is fine, so long as we properly address changes to the `minCapacity` pool config. We could err on the side of caution and put it behind a feature flag at first to only enable on certain pools for testing/validating, then later drop it, and make that behavior the default for `minCapacity`.

JohanLorenzo replied:
Yeah, I believe we should make the behavior the default one once we're confident enough with the change. So, I'd be fine to `minCapacityIdleTimeoutSecs` whenever we don't need it anymore.

JohanLorenzo replied:
Closing thread in favor of https://github.com/taskcluster/taskcluster-rfcs/pull/192#discussion_r2406155765: `minCapacityIdleTimeoutSecs` was replaced by a boolean flag and that flag is questioned as well. So let's continue the conversation there.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+
+**Worker Provisioning Logic:**
+- Worker-manager determines worker idle timeout configuration at spawn time based on current pool capacity
+- Uses existing [`launch_config_id` system](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/data.js#L313-L362) to ensure per-variant capacity tracking
```
I think we can drop this line or maybe rephrase it a bit? 
I remember there was bug at some point where setting odd `minCapacity` on a pool that has `capacityPerInstance > 1` was leading to some "interesting" behavior

matt-boris replied:
> I remember there was bug at some point where setting odd `minCapacity` on a pool that has `capacityPerInstance > 1` was leading to some "interesting" behavior

Yes, this was on docker worker pools that can have worker capacity > 1

JohanLorenzo replied:
Line dropped.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+
+#### 2. Lifecycle-Based Worker Configuration
+
+Workers are assigned `minCapacity` or overflow roles at launch time and never change configuration during their lifetime. This approach works within TaskCluster's architectural constraints where worker-manager cannot communicate configuration changes to running workers.
```
maybe we could keep it simple with a single boolean, we only care if the worker is launched as min capacity (forever-worker) or just a regular
`is_min_capacity` or `is_going_to_work_forever` 😃

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+
+**Validation:**
+- Fail pool provisioning entirely if invalid configuration (e.g., `minCapacity > maxCapacity`)
+- Require `minCapacityIdleTimeoutSecs >= 0` if `minCapacity` workers are desired
```
so are we going to allow any values in here? should it be then also checked if `minCapacityIdleTimeoutSecs` is lower than the regular `idleTimeoutSecs`? 
Or we only allow `0` as a value

JohanLorenzo replied:
Let's use a binary flag instead of allowing any value. That'll make the implementation simpler.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+
+**Capacity Management:**
+- **minCapacity increases**: Spawn additional `minCapacity` workers with indefinite runtime
+- **minCapacity decreases**: Forcefully terminate excess `minCapacity` workers (see Capacity Reduction Trade-offs)
```
this will probably depend on the `minCapacityIdleTimeoutSecs` if used. For some numbers it might be better not to do anything at all

JohanLorenzo replied:
`minCapacityIdleTimeoutSecs` won't be used in favor of a binary flag.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+When `minCapacity` decreases, excess `minCapacity` workers must be forcefully terminated since:
+- Worker-manager cannot communicate with running workers to reduce their idle timeouts
+- Waiting for natural idle timeout could take hours (defeating responsive capacity management)
+- Forceful termination provides immediate capacity adjustment but loses cache benefits
```
whenever we kill a running worker there is a high chance of getting claim expired, as he might be picking up new tasks at the same time. I wonder if we should treat those in some special way. I.e. mark them as "quarantined for removal" to make sure they won't pick new tasks, and only then remove (maybe on the next scanner run)

JohanLorenzo replied:
Great call! I changed the RFC to first quarantine the worker.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+- ❌ **Cache loss**: Defeats primary optimization goal when terminating active workers
+- ⚠️ **Selective termination**: Prioritize terminating idle workers when possible
+
+#### 6. Enhanced Health Monitoring
```
I don't understand this section. 
What does "more agressive health monitoring" means? Or "faster detection of unresponsive workers"

JohanLorenzo replied:
Section removed given that it was causing more confusion than anything.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+**Implementation Location:**
+New metrics will be added to [services/worker-manager/src/monitor.js](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/monitor.js#L377) following the existing pattern established in [lines 246-377](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/monitor.js#L246- ...
+
+**New Prometheus Metrics:**
```
Do we really need all of those metrics? 
I feel like existing metrics are enough, and in case we stumble upon some errors we'd be looking in longs anyway. 
We have `*_queue_running_workers` to count how many active workers there are

JohanLorenzo replied:
New metrics removed in favor of existing ones. Thanks!

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+**Metrics Integration:**
+Metrics will be exposed in [estimator.js](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/estimator.js#L89-L92) using the existing `exposeMetrics()` pattern to report worker counts by pool and role during each provisioning cycle.
+
+**Alerting Rules:**
```
I don't think this alert would be actionable, maybe we don't need this?

JohanLorenzo replied:
Alert removed!

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+## Error Handling and Edge Cases
+
+**Worker Lifecycle Management:**
+- **Pool reconfiguration**: Capacity changes trigger worker replacement, not reconfiguration
```
I think this is a very important edge-case for the long-running (or forever running workers)
If launch config is changed /removed/ etc (either for cost reasons, or if we want to upgrade instance type or image itself) what do we do? 

My intuition says if launch config is being archived (not present in new config) and there are long running workers created from it, workers should be killed

JohanLorenzo replied:
I added a section to cover this edge case. Let me know what you think!

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+**Capacity Reduction Trade-offs:**
+When `minCapacity` decreases, excess `minCapacity` workers must be forcefully terminated since:
+- Worker-manager cannot communicate with running workers to reduce their idle timeouts
+- Waiting for natural idle timeout could take hours (defeating responsive capacity management)
```
does it mean that this proposal encourages worker manager to just kill those workers even when it might be useful to execute pending tasks?

JohanLorenzo replied:
Fixed by quarantining workers first.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+- **Azure**: [Worker ID maps to VM name](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/providers/azure/index.js#L302), uses [VM deletion API](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/providers/azure/index.js#L1249-L1254)
+
+**Implementation:**
+Worker-manager has existing capability to forcefully terminate workers by updating the database to mark workers as STOPPING state and then making direct cloud provider API calls to terminate the instances. This process works across all supported cloud providers (GCP, AWS, Azure) using their respective termination APIs.
```
I think we should be specific in RFC how and when we want to "get rid" of unwanted workers. 
If scale down happens, ideally we want to make sure those workers are not running tasks when we are trying to get rid of them.

Existing worker scanner loop goes pool by pool and checks all workers. 
It could count all "minCapacity" workers and as soon it reaches `> wp.minCapacity` value, it  could mark those as "excess". For example we have `quarantineWorker({ reason })` functionality that could mark worker as quarantined with the attached reason `minCapacity reduction` so we could guarantee it will not be given new task. 

And during next scan loop it could simply check if is: 1 - minCapacity worker, 2 - was quarantined, 3 - not running any task ==> safe to be force deleted.

Or maybe that's an over-engineering, and we just remove workers on the first scan

JohanLorenzo replied:
Good idea! I updated the RFC to delete `minCapacity` workers in 2 phases: 
 1. Quarantining it
 2. Terminating it.

Let me know if that matches what you have in mind!

matt-boris replied:
There'd need to be a check to ensure the worker isn't currently working on any task, as the quarantine could happen while it's doing work. We don't want to terminate underneath a task that's being executed.

JohanLorenzo replied:
Sorry, my message didn't say that but waiting for the task to finish is now part of the RFC 🙂 

https://github.com/taskcluster/taskcluster-rfcs/pull/192/files#diff-715f49575df52b5b8c0fd5a04a69747bb67d486f64659b458dd3ab6776771ca6R114

lotas replied:
I'm pretty sure quarantine feature only prevents worker from claiming new tasks, but won't affect existing running task, so complete/fail should still be fine

JohanLorenzo replied:
Great! Resolving thread, then 🙂

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,229 @@
+**Alerting Rules:**
+Two key alerting rules will monitor `minCapacity` worker health: a MinCapacityDeficit alert triggered when worker pools fall below `minCapacity` for more than 30 seconds, and a HighMinCapacityTerminations alert triggered when termination rates exceed 0.1 workers per 5-minute window for more than 2 minutes.
+
+## Success Measurement
```
do we want to keep it as part of an RFC? I personally feel like it shouldn't stay in the RFC itself, but can be given as a reason why we want to have this feature. idk :)

JohanLorenzo replied:
Section removed.

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,199 @@
+#### 3. Pool Configuration
+
+**New Configuration Options:**
+Pools will have a new boolean flag `enableMinCapacityWorkers` that enables minCapacity worker behavior. When enabled, workers fulfilling minCapacity requirements will run indefinitely (idle timeout set to 0), providing maximum cache preservation.
```
I wonder if we really want an extra flag for this. For me `minCapacity` is already there. 
Unless someone wanted to use it the way they were currently working, I think we can just not introduce a new flag. 
just  thinking out loud

JohanLorenzo replied:
I was thinking of a new flag in order to roll out the changes. But maybe there's another way I didn't think of. I'm open to other options 🙂

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,199 @@
+
+**Phase 1 - Quarantine (First Scanner Run):**
+1. Worker-manager counts all minCapacity workers in the pool
+2. When the count exceeds the pool's `minCapacity` setting, identify excess workers
```
I wonder if we should explain how we would pick up those candidates.. maybe just killing the oldest one first?

JohanLorenzo replied:
Good idea. Done.

JohanLorenzo commented:
@jcristau, I remember you had some thoughts on the current behavior. What do you think of this proposal?

lotas commented:
If I understand correctly @petemoore,  you suggest:

1. Add a feature to g-w (worker-runner) to ask worker-manager if it should keep on doing work or should terminate itself (as soon as appropriate)
2. let worker-manager scale down automatic pools. 

Aren't they competing solutions? 

I like the idea of making worker manager not aware of worker configs, probably less error-prone in the long run 👍 

I think there could be edge cases on both sides of those solutions, for example, if we make all workers long-lived ones, and worker-manager fails in some way to provision/scan (like expired credentials or internal error), which would result in many workers running way longer than they should. 

Johan's initial proposal is probably a good temporary fix to let some workers live longer, but implementing some sort of signaling mechanism where workers talk to worker manager to know if they should keep going instead of shutting down might be better in the long run.

petemoore commented:
I wrote a lot of words, but in summary, I think I’m suggesting the approach is ok in principle, but should be for the whole pool, instead of splitting it in two.

This way, Worker Manager is free to decide which instances to terminate, based on its own conditions, which may change over time.

With Worker Manager managing the termination of workers, we can either turn off the feature on the workers directly in the Generic-Worker config, or have some longer fallback period, to catch situations like the one you highlight.

Regarding how Worker Manager communicates with the worker to ask it to gracefully shutdown, I am reasonably agnostic. Perhaps the quarantine followed by a termination isn't the worst thing, although it doesn't really allow workers to perform any cleanup (e.g. imagine in future that workers might want to e.g. preserve their caches by writing them to a storage bucket before terminating, or securely erasing secrets, or writing an activity report somewhere, reporting lifecycle events to some audit trail, or any other teardown type activity that might be useful...).

In any case, I think any time Worker Manager decides it no longer requires an instance, there should be some way to signal it to a worker, so that it can gracefully terminate, but Worker Manager should keep track of the instance and forcefully kill it after some timeout, if the worker hasn't successfully terminated itself.

lotas commented:
Worker manager already can `removeWorker()` which will tell instance to shutdown, and worker-runner/generic-worker would get notified when this happens. But I guess it comes without any warning to the instance that is about to shutdown - something like spot termination is doing where you can potentially upload stuff before shutting down. 

If there was some similar way to send `SIGTERM` signal to a vm and let g-w wrap things up before getting hard kill event that would make it easier probably

JohanLorenzo commented:
Thank you for your input, @petemoore! I agree having a central place like worker-manager makes more sense and avoid duplication of responsibilities 👍  I just added a phase 2 to cover that. For 2025, I'd like us to focus on phase 1, which is leveraging `idleTimeoutSecs=0` 

>  For example, prices might drop for one instance type, but you have workers set to run forever with a different instance type, so you are stuck with the ones you have.

For the record, this use case doesn't exist in fxci. `yq '.pools[] | select((.config.instance_types | length) > 1) | .pool_id' worker-pools.yml` doesn't return anything. 

In community, there are just 2 pools: `git-cinnabar/windows` and `relman/win2022`. The former is not supported anymore now that Firefox moved to https://github.com/mozilla-firefox/firefox. The latter is then the only exception. I believe it'd be better to split this pool per instance type rather than trying to find a design because of it. What do you think? `yq` command for community.

```
yq '
    (filename | sub(".*/", "") | sub(".yml$", "")) as $file |
    keys[] as $proj |
    .[$proj].workerPools // {} |
    to_entries[] |
    select((.value.instanceTypes // {} | length) > 1) |
    $file + ": " + $proj + "/" + .key
  ' config/projects/*.yml
```


> This way we do not muddy the Quarantine logs with events that are for terminations, not for quarantines

TIL we have specific quarantine logs. Where I can find them? 

> Consider also that technically a worker can perform work for a worker pool even if it wasn't launched by Worker Manager.

Do we have a way to measure how often this happens? I'm under the impression that it rarely occurs on Firefox CI because spawning a new instance is too complex and most people don't have enough permissions to spawn their own worker. I might be wrong though. I'd love to see if what it's really like.

petemoore commented:
> Thank you for your input, @petemoore! I agree having a central place like worker-manager makes more sense and avoid duplication of responsibilities 👍 I just added a phase 2 to cover that. For 2025, I'd like us to focus on phase 1, which is leveraging `idleTimeoutSecs=0`
> 
> > For example, prices might drop for one instance type, but you have workers set to run forever with a different instance type, so you are stuck with the ones you have.
> 
> For the record, this use case doesn't exist in fxci. `yq '.pools[] | select((.config.instance_types | length) > 1) | .pool_id' worker-pools.yml` doesn't return anything.
> 
> In community, there are just 2 pools: `git-cinnabar/windows` and `relman/win2022`. The former is not supported anymore now that Firefox moved to https://github.com/mozilla-firefox/firefox. The latter is then the only exception. I believe it'd be better to split this pool per instance type rather than trying to find a design because of it. What do you think? `yq` command for community.
> 
> ```
> yq '
>     (filename | sub(".*/", "") | sub(".yml$", "")) as $file |
>     keys[] as $proj |
>     .[$proj].workerPools // {} |
>     to_entries[] |
>     select((.value.instanceTypes // {} | length) > 1) |
>     $file + ": " + $proj + "/" + .key
>   ' config/projects/*.yml
> ```

This was just an _example_ of one scenario where Worker Manager at runtime may have reason to prefer keeping one worker alive over another, from information not available at the time it launched the "run forever" instance. The point being, if the a matter is decided in worker config, it is locked in and non-changeable as soon as the instance is spawned.

> 
> > This way we do not muddy the Quarantine logs with events that are for terminations, not for quarantines
> 
> TIL we have specific quarantine logs. Where I can find them?

I'm not sure if we have quarantine logs per se, but the Queue provides an API to query quarantine history.

> 
> > Consider also that technically a worker can perform work for a worker pool even if it wasn't launched by Worker Manager.
> 
> Do we have a way to measure how often this happens? I'm under the impression that it rarely occurs on Firefox CI because spawning a new instance is too complex and most people don't have enough permissions to spawn their own worker. I might be wrong though. I'd love to see if what it's really like.

Yes I doubt it often happens, it was more an observation about the limitations of the proposed design. There is no technical restriction from being able to supplement a Worker Manager worker pool with additional e.g. hardware workers. Now imagine the minimum pool size is set to 5 and Worker Manager starts up, and creates 5 live-forever workers. At some point, someone decides it is better for the 5 workers that live forever to be hardware workers, for whatever reason. They create 5 hardware workers that live forever. Worker Manager now sees there are 10 workers running, but no tasks, and the minimum worker pool size is 5. So ideally it would terminate the 5 it started up. Of course, an operations staff member can also delete them manually, so no big deal. It is just one extra thing for a person to mentally track.

However, given that the design as proposed is relatively straightforward to implement, maybe we can go with that for now. Worker Manager should not need to know about Generic Worker config settings, but it is a hack we can live with, if we can agree to remove it in the future. It really should be part of the protocol of interaction between worker and worker manager.

Maybe a slightly cleaner solution would be when the worker registers, it gets a worker manager defined parameter about whether it should stay alive forever, or what its idle timeout should be. At this point, worker runner or the worker can translate it into its own config setting.

JohanLorenzo commented:
Agreed. Let's go with the proposed design for now and revisit the day we decide to mix hardware workers among cloud-based ones.

jcristau commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+   - Package manager caches (npm, pip, cargo, etc.)
+   - Container images and layers
+
+2. **Provisioning Delays**: New worker provisioning [takes ~75 seconds average for decision pools](https://taskcluster.github.io/mozilla-history/worker-metrics), during which tasks must wait
```
IMO the average is not the issue here, it's that whenever something goes slightly wrong, or it's a slightly less often used pool, it can take tens of minutes.

JohanLorenzo commented:
Discussed with the whole Taskcluster team at today's community call. We agreed the RFC is in good shape for entering the Final Comment phase.

jcristau commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+- **Resource allocation**: minCapacity workers mixed with other workers on same infrastructure
+
+**Launch Configuration Changes:**
+When a launch configuration is changed, removed, or archived, all workers created from the old configuration must be terminated and replaced:
+- If a launch configuration is archived (not present in new configuration), identify all long-running workers created from it
+- Terminate these workers via cloud provider APIs after checking for running tasks
```
How does that work?  Do we have a way to signal to the worker that it shouldn't claim any more tasks?  Or signal to the queue that it shouldn't give it any?  Or do we risk killing a running task?

JohanLorenzo replied:
Worker-manager is able to terminate cloud instances ([e.g.](
https://github.com/taskcluster/taskcluster/blob/754938c53ba34aea5a50ce610272e7a275c11911/services/worker-manager/src/providers/google.js#L407))

In its current state, the RFC risks killing a running task. In a previous revision, we thought of quarantining the worker. However @petemoore [raised the concern](https://github.com/taskcluster/taskcluster-rfcs/pull/192#pullrequestreview-3410503724) that it would bend the purpose of the quarantine mechanism.

lotas replied:
So I think what we are missing here is the guarantee that worker is not running any tasks and the queue will not give out new tasks.
Maybe we need some extra mechanism here in addition to quarantine mechanism, that would signal queue to mark worker as "scheduled for removal" which would essentially do the same but have clear signal that we are about to decommission this worker. 

Could be a short-lived queue (~20min same as claim timeout)

jcristau commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+
+Worker-manager enforces idle timeout using existing [`queueInactivityTimeout` from lifecycle configuration](https://github.com/taskcluster/taskcluster/blob/754938c53ba34aea5a50ce610272e7a275c11911/services/worker-manager/schemas/v1/worker-lifecycle.yml#L33-L50) and through idleTimeout (as specified in phase 1).
+
+[Scanner polls](https://github.com/taskcluster/taskcluster/blob/754938c53ba34aea5a50ce610272e7a275c11911/services/worker-manager/src/worker-scanner.js#L69-L76) Queue API to track worker idle time:
```
Do we expect this to scale well?  (We already have issues with scanner loops taking too long.)

JohanLorenzo replied:
@lotas, what do you think?

lotas replied:
yeah, we probably don't want to kill queue api with those increased loads of queries. 

Instead, I think we should build some state inside w-m service. 
One idea could be to listen to queue task related events and update some form of cache or state with the following details:
- `workerId`
- `lastTaskId`
- `lastTaskStartedAt`
- `lastTaskResolvedAt`

this way we'll simply leverage existing rabbitmq exchanges, and would know the approximate state of the worker without ddosing queue service

jcristau commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+
+All terminations check for running tasks before proceeding.
+
+**Migration:**
+Deploy as breaking change requiring simultaneous worker-manager and generic-worker updates.
```
This seems like a fairly risky strategy, best avoided...

JohanLorenzo replied:
@petemoore, could we have a way for generic-worker to support both idle-termination and termination from worker-manager for a transition period?

jcristau commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+- Launch configuration changed/archived
+- Worker is unhealthy (provider-specific check)
+
+All terminations check for running tasks before proceeding.
```
As in phase 1, care will be needed here to avoid races between termination and claimWork, it would be good to flesh out that process more.

JohanLorenzo replied:
@jcristau What would you recommend?

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+
+**Removing Excess Capacity:**
+
+When `runningCapacity > minCapacity`, [worker-manager scanner identifies](https://github.com/taskcluster/taskcluster/blob/754938c53ba34aea5a50ce610272e7a275c11911/services/worker-manager/src/worker-scanner.js#L69-L76) and terminates excess workers.
```
As it currently stands, worker-scanner doesn't remove workers when capacity changes, it only updates db state. 

Are we talking here about the case when `minCapacity` was lowered and we need to reduce the number of minCapacity-workers provisioned earlier?

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+When `runningCapacity > minCapacity`, [worker-manager scanner identifies](https://github.com/taskcluster/taskcluster/blob/754938c53ba34aea5a50ce610272e7a275c11911/services/worker-manager/src/worker-scanner.js#L69-L76) and terminates excess workers.
+
+**Termination logic:**
+- Query [Queue API client](https://github.com/taskcluster/taskcluster/blob/754938c53ba34aea5a50ce610272e7a275c11911/services/worker-manager/src/main.js#L176-L178) to check if worker's latest task is running
```
this will likely not guarantee anything (workers currently polling for work), but would increase load on the queue service api.

Assuming we want to implement this active workers termination routines already, I would propose to keep it really simple, and only select the desired amount of oldest workers to be terminated

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+
+**Launch Configuration Changes:**
+When a launch configuration is changed, removed, or archived, all workers created from the old configuration must be terminated and replaced:
+- If a launch configuration is archived (not present in new configuration), identify all long-running workers created from it
```
this is actually a good point, and probably should be solved separately. Launch configs will be archived if something changes, and it's a good question if they should be killed immediately or not. Because if there was a minor change, like zone removed, do we want to let it work until it idles and shuts down, or we want to kill all such workers right away

lotas commented on rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md:
```diff
@@ -0,0 +1,133 @@
+Worker-manager terminates workers when:
+- Idle timeout exceeded (`queueInactivityTimeout`)
+- Capacity exceeds `maxCapacity`
+- Capacity exceeds `minCapacity` (terminate oldest first)
```
wait, wouldn't it contradict the whole point of min capacity workers? If we'll always be killing oldest, those with caches would be gone first

JohanLorenzo commented:
Discussed in person with @lotas, @petemoore, @hneiva. We decided to take a different and more generic approach. We decided to close this current RFC and to create a new one with the new proposal. The main reason for this is: the new approach requires a full rewrite of the RFC, so the comments in this current RFC will create more noise than anything.

ahal commented:
@lotas can this be closed?

lotas commented:
> @lotas can this be closed?

indeed, thanks for reminder @ahal 

This was implemented slightly differently in form of "Should worker terminate": 
https://github.com/taskcluster/taskcluster/pull/8112
https://github.com/taskcluster/taskcluster/pull/8282
(and few more)
